File: src/main/java/com/sems/devices/interfaces/rest/resources/DeviceResources.java
package com.sems.devices.interfaces.rest.resources;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.sems.devices.domain.model.aggregates.Device;
import com.sems.devices.domain.model.entities.DeviceBinding;
import com.sems.devices.domain.model.entities.DeviceConfiguration;
import com.sems.devices.domain.model.entities.DeviceEvent;
import jakarta.validation.constraints.NotBlank;
import java.time.Instant;

/**
 * Contrato JSON del modulo de dispositivos.
 *
 * <p>Los nombres de campo replican exactamente los del servicio en Go, incluido
 * el uso de camelCase y la omision de los opcionales vacios. El frontend no
 * distingue si detras hay Go o Java.
 */
public final class DeviceResources {

    private DeviceResources() {
    }

    // ------------------------------------------------------------- peticiones

    public record CreateDeviceRequest(
            @NotBlank(message = "is required") String externalDeviceCode,
            @NotBlank(message = "is required") String userId,
            @NotBlank(message = "is required") String deviceName,
            @NotBlank(message = "is required") String deviceType,
            String brand,
            String model,
            @NotBlank(message = "is required") String connectionProtocol) {
    }

    public record UpdateDeviceRequest(
            @NotBlank(message = "is required") String deviceName,
            @NotBlank(message = "is required") String deviceType,
            String brand,
            String model,
            @NotBlank(message = "is required") String connectionProtocol) {
    }

    public record UpdateDeviceStatusRequest(
            @NotBlank(message = "is required") String status) {
    }

    public record CreateBindingRequest(
            @NotBlank(message = "is required") String userId,
            String homeId) {
    }

    public record CreateConfigurationRequest(
            @NotBlank(message = "is required") String configKey,
            String configValue) {
    }

    public record UpdateConfigurationRequest(String configValue) {
    }

    public record CreateEventRequest(
            @NotBlank(message = "is required") String eventType,
            String description,
            Instant occurredAt) {
    }

    // -------------------------------------------------------------- respuestas

    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record DeviceResource(
            String deviceId, String externalDeviceCode, String userId, String deviceName,
            String deviceType, String brand, String model, String connectionProtocol,
            String status, Instant registeredAt, Instant updatedAt) {

        public static DeviceResource from(Device d) {
            return new DeviceResource(
                    d.getDeviceId().toString(), d.getExternalDeviceCode(), d.getUserId().toString(),
                    d.getDeviceName(), d.getDeviceType(), d.getBrand(), d.getModel(),
                    d.getConnectionProtocol().name(), d.getStatus().name(),
                    d.getRegisteredAt(), d.getUpdatedAt());
        }
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record DeviceBindingResource(
            String bindingId, String deviceId, String userId, String homeId,
            String bindingStatus, Instant linkedAt, Instant unlinkedAt, Instant updatedAt) {

        public static DeviceBindingResource from(DeviceBinding b) {
            return new DeviceBindingResource(
                    b.getBindingId().toString(), b.getDeviceId().toString(), b.getUserId().toString(),
                    b.getHomeId() == null ? null : b.getHomeId().toString(),
                    b.getBindingStatus().name(), b.getLinkedAt(), b.getUnlinkedAt(), b.getUpdatedAt());
        }
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record DeviceConfigurationResource(
            String configurationId, String deviceId, String configKey,
            String configValue, Instant updatedAt) {

        public static DeviceConfigurationResource from(DeviceConfiguration c) {
            return new DeviceConfigurationResource(
                    c.getConfigurationId().toString(), c.getDeviceId().toString(),
                    c.getConfigKey(), c.getConfigValue(), c.getUpdatedAt());
        }
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record DeviceEventResource(
            String eventId, String deviceId, String eventType,
            String description, Instant occurredAt) {

        public static DeviceEventResource from(DeviceEvent e) {
            return new DeviceEventResource(
                    e.getEventId().toString(), e.getDeviceId().toString(),
                    e.getEventType(), e.getDescription(), e.getOccurredAt());
        }
    }
}

File: src/Sems.Api/Modules/Devices/Interfaces/DeviceResources.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;
using Sems.Api.Modules.Devices.Domain.Model;

namespace Sems.Api.Modules.Devices.Interfaces;

/// <summary>
/// Contrato JSON del modulo de dispositivos.
///
/// <para>Los nombres replican exactamente los del servicio en Go, en camelCase.
/// El frontend no distingue si detras hay Go, Java o C#.</para>
/// </summary>
public static class DeviceResources
{
    // ------------------------------------------------------------- peticiones

    public sealed record CreateDeviceRequest(
        [Required(ErrorMessage = "is required")] string ExternalDeviceCode,
        [Required(ErrorMessage = "is required")] string UserId,
        [Required(ErrorMessage = "is required")] string DeviceName,
        [Required(ErrorMessage = "is required")] string DeviceType,
        string? Brand,
        string? Model,
        [Required(ErrorMessage = "is required")] string ConnectionProtocol);

    public sealed record UpdateDeviceRequest(
        [Required(ErrorMessage = "is required")] string DeviceName,
        [Required(ErrorMessage = "is required")] string DeviceType,
        string? Brand,
        string? Model,
        [Required(ErrorMessage = "is required")] string ConnectionProtocol);

    public sealed record UpdateDeviceStatusRequest(
        [Required(ErrorMessage = "is required")] string Status);

    public sealed record CreateBindingRequest(
        [Required(ErrorMessage = "is required")] string UserId,
        string? HomeId);

    public sealed record CreateConfigurationRequest(
        [Required(ErrorMessage = "is required")] string ConfigKey,
        string? ConfigValue);

    public sealed record UpdateConfigurationRequest(string? ConfigValue);

    public sealed record CreateEventRequest(
        [Required(ErrorMessage = "is required")] string EventType,
        string? Description,
        DateTime? OccurredAt);

    // -------------------------------------------------------------- respuestas

    public sealed record DeviceResource(
        string DeviceId, string ExternalDeviceCode, string UserId, string DeviceName,
        string DeviceType,
        [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] string? Brand,
        [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] string? Model,
        string ConnectionProtocol, string Status, DateTime RegisteredAt, DateTime UpdatedAt)
    {
        public static DeviceResource From(Device d) => new(
            d.DeviceId.ToString(), d.ExternalDeviceCode, d.UserId.ToString(), d.DeviceName,
            d.DeviceType, d.Brand, d.Model, d.ConnectionProtocol.ToString(), d.Status.ToString(),
            d.RegisteredAt, d.UpdatedAt);
    }

    public sealed record DeviceBindingResource(
        string BindingId, string DeviceId, string UserId,
        [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] string? HomeId,
        string BindingStatus, DateTime LinkedAt,
        [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] DateTime? UnlinkedAt,
        DateTime UpdatedAt)
    {
        public static DeviceBindingResource From(DeviceBinding b) => new(
            b.BindingId.ToString(), b.DeviceId.ToString(), b.UserId.ToString(),
            b.HomeId?.ToString(), b.BindingStatus.ToString(), b.LinkedAt, b.UnlinkedAt, b.UpdatedAt);
    }

    public sealed record DeviceConfigurationResource(
        string ConfigurationId, string DeviceId, string ConfigKey,
        [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] string? ConfigValue,
        DateTime UpdatedAt)
    {
        public static DeviceConfigurationResource From(DeviceConfiguration c) => new(
            c.ConfigurationId.ToString(), c.DeviceId.ToString(), c.ConfigKey, c.ConfigValue,
            c.UpdatedAt);
    }

    public sealed record DeviceEventResource(
        string EventId, string DeviceId, string EventType,
        [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] string? Description,
        DateTime OccurredAt)
    {
        public static DeviceEventResource From(DeviceEvent e) => new(
            e.EventId.ToString(), e.DeviceId.ToString(), e.EventType, e.Description, e.OccurredAt);
    }
}
